plugins/modules/ec2_eip.py

```diff
@@ -249,110 +264,127 @@ def associate_ip_and_device(
                 )
                 if private_ip_address:
                     params["PrivateIpAddress"] = private_ip_address
-                if address["Domain"] == "vpc":
-                    params["AllocationId"] = address["AllocationId"]
+                if address.get("Domain") == "vpc":
+                    params["AllocationId"] = address.get("AllocationId")
                 else:
-                    params["PublicIp"] = address["PublicIp"]
-                res = ec2.associate_address(aws_retry=True, **params)
-            except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+                    params["PublicIp"] = address.get("PublicIp")
+                res = associate_address(ec2, **params)
+            except AnsibleEC2Error as e:
                 msg = f"Couldn't associate Elastic IP address with instance '{device_id}'"
                 module.fail_json_aws(e, msg=msg)
         else:
             params = dict(
                 NetworkInterfaceId=device_id,
-                AllocationId=address["AllocationId"],
+                AllocationId=address.get("AllocationId"),
                 AllowReassociation=allow_reassociation,
             )
 
             if private_ip_address:
                 params["PrivateIpAddress"] = private_ip_address
 
             try:
-                res = ec2.associate_address(aws_retry=True, **params)
-            except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+                res = associate_address(ec2, **params)
+            except AnsibleEC2Error as e:
                 msg = f"Couldn't associate Elastic IP address with network interface '{device_id}'"
                 module.fail_json_aws(e, msg=msg)
         if not res:
             module.fail_json(msg="Association failed.")
 
-    return {"changed": True}
+    return True
 
 
-def disassociate_ip_and_device(ec2, module, address, device_id, check_mode, is_instance=True):
+def disassociate_ip_and_device(
+    ec2,
+    module: AnsibleAWSModule,
+    address: Optional[Dict[str, Union[str, List[Dict[str, str]]]]],
+    device_id: str,
+    is_instance: bool = True,
+) -> bool:
     if not address_is_associated_with_device(ec2, module, address, device_id, is_instance):
-        return {"changed": False}
+        return False
 
     # If we're in check mode, nothing else to do
-    if not check_mode:
+    if not module.check_mode:
+        if not address:
```

Comment:
What do you think about modifying line 632 to something like?

       if device_id:
            address = find_address(ec2, module, public_ip, device_id, is_instance=is_instance) or {}
        else:
            address = find_address(ec2, module, public_ip, None) or {}

and limit these checks?
```
        if not address:
            address = {}
```

Reply:
I have done some more code refactoring, could you please review once again? Thanks